src/Nethermind/Nethermind.Evm/DataGasCalculator.cs

```diff
@@ -0,0 +1,108 @@
+// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using Nethermind.Core;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+
+namespace Nethermind.Evm;
+
+public static class DataGasCalculator
+{
+    public static ulong CalculateDataGas(int blobCount) =>
+        (ulong)blobCount * Eip4844Constants.DataGasPerBlob;
+
+    public static ulong CalculateDataGas(Transaction transaction) =>
+        CalculateDataGas(transaction.BlobVersionedHashes?.Length ?? 0);
+
+    public static ulong CalculateDataGas(Transaction[] transactions)
+    {
+        int blobCount = 0;
+        foreach (Transaction tx in transactions)
+        {
+            if (tx.SupportsBlobs)
+            {
+                blobCount += tx.BlobVersionedHashes!.Length;
+            }
+        }
+
+        return CalculateDataGas(blobCount);
+    }
+
+    public static bool TryCalculateDataGasPrice(BlockHeader header, Transaction transaction, out UInt256 dataGasPrice)
+    {
+        if (!TryCalculateDataGasPricePerUnit(header.ExcessDataGas.Value, out UInt256 dataGasPricePerUnit))
+        {
+            dataGasPrice = UInt256.MaxValue;
+            return false;
+        }
+        return !UInt256.MultiplyOverflow(CalculateDataGas(transaction), dataGasPricePerUnit, out dataGasPrice);
+    }
+
+    public static bool TryCalculateDataGasPricePerUnit(BlockHeader header, out UInt256 dataGasPricePerUnit) =>
+        header.ExcessDataGas is null
+            ? throw new ArgumentException(nameof(BlockHeader.ExcessDataGas))
+            : TryCalculateDataGasPricePerUnit(header.ExcessDataGas.Value, out dataGasPricePerUnit);
+
+    public static bool TryCalculateDataGasPricePerUnit(ulong excessDataGas, out UInt256 dataGasPricePerUnit)
+    {
+        static bool FakeExponentialOverflow(UInt256 factor, UInt256 num, UInt256 denominator, out UInt256 dataGasPricePerUnit)
```

Comment:
This doesn't look optimal?
Does this belong in some kind of UInt256 extensions class or UInt256 itself?